File: Sources/CodexManager/Behavior/SettingsCoordinator.swift
import Foundation

actor SettingsCoordinator {
    private let settingsRepository: SettingsRepository
    private let launchAtStartupService: LaunchAtStartupServiceProtocol

    init(
        settingsRepository: SettingsRepository,
        launchAtStartupService: LaunchAtStartupServiceProtocol
    ) {
        self.settingsRepository = settingsRepository
        self.launchAtStartupService = launchAtStartupService
    }

    func currentSettings() throws -> AppSettings {
        try settingsRepository.loadSettings()
    }

    func updateSettings(_ patch: AppSettingsPatch) throws -> AppSettings {
        let launchAtStartupPatch = patch.launchAtStartup

        var settings = try settingsRepository.loadSettings()

        if let value = patch.launchAtStartup { settings.launchAtStartup = value }
        if let value = patch.launchCodexAfterSwitch { settings.launchCodexAfterSwitch = value }
        if let value = patch.autoSmartSwitch { settings.autoSmartSwitch = value }
        if let value = patch.syncOpencodeOpenaiAuth { settings.syncOpencodeOpenaiAuth = value }
        if let value = patch.localProxyHostAPIOnly { settings.localProxyHostAPIOnly = value }
        if let value = patch.restartEditorsOnSwitch { settings.restartEditorsOnSwitch = value }
        if let value = patch.restartEditorTargets { settings.restartEditorTargets = value }
        if let value = patch.autoStartApiProxy { settings.autoStartApiProxy = value }
        if let value = patch.proxyConfiguration { settings.proxyConfiguration = value.normalized() }
        if let value = patch.remoteServers { settings.remoteServers = value }
        if let value = patch.locale { settings.locale = AppLocale.resolve(value).identifier }

        try settingsRepository.saveSettings(settings)

        if let launchAtStartupPatch {
            try launchAtStartupService.setEnabled(launchAtStartupPatch)
        }

        return settings
    }

    func syncLaunchAtStartupFromStore() throws {
        let settings = try settingsRepository.loadSettings()
        try launchAtStartupService.syncWithStoreValue(settings.launchAtStartup)
    }
}

File: apps/desktop/src/main/services/settings-coordinator.ts
import type { AppSettings, AppSettingsPatch } from "../../shared/models/settings";
import { normalizeEditorAppIDs, resolveAppLocale } from "../../shared/models/settings";

export interface SettingsRepositoryLike {
  loadSettings(): Promise<AppSettings>;
  saveSettings(settings: AppSettings): Promise<void>;
}

export interface LaunchAtStartupServiceLike {
  setEnabled(enabled: boolean): void | Promise<void>;
  syncWithStoreValue(enabled: boolean): void | Promise<void>;
}

export class SettingsCoordinator {
  constructor(
    private readonly settingsRepository: SettingsRepositoryLike,
    private readonly launchAtStartupService: LaunchAtStartupServiceLike
  ) {}

  async currentSettings(): Promise<AppSettings> {
    return this.settingsRepository.loadSettings();
  }

  async updateSettings(patch: AppSettingsPatch): Promise<AppSettings> {
    const launchAtStartupPatch = patch.launchAtStartup;
    const settings = await this.settingsRepository.loadSettings();

    if (patch.launchAtStartup !== undefined) {
      settings.launchAtStartup = patch.launchAtStartup;
    }
    if (patch.launchCodexAfterSwitch !== undefined) {
      settings.launchCodexAfterSwitch = patch.launchCodexAfterSwitch;
    }
    if (patch.autoSmartSwitch !== undefined) {
      settings.autoSmartSwitch = patch.autoSmartSwitch;
    }
    if (patch.restartEditorsOnSwitch !== undefined) {
      settings.restartEditorsOnSwitch = patch.restartEditorsOnSwitch;
    }
    if (patch.restartEditorTargets !== undefined) {
      settings.restartEditorTargets = normalizeEditorAppIDs(patch.restartEditorTargets);
    }
    if (patch.locale !== undefined) {
      settings.locale = resolveAppLocale(patch.locale);
    }
    if (patch.proxyPort !== undefined) {
      settings.proxyPort = patch.proxyPort;
    }
    if (patch.proxyApiKey !== undefined) {
      settings.proxyApiKey = patch.proxyApiKey;
    }
    if (patch.autoStartProxy !== undefined) {
      settings.autoStartProxy = patch.autoStartProxy;
    }

    await this.settingsRepository.saveSettings(settings);
    if (launchAtStartupPatch !== undefined) {
      await this.launchAtStartupService.setEnabled(launchAtStartupPatch);
    }
    return settings;
  }

  async syncLaunchAtStartupFromStore(): Promise<void> {
    const settings = await this.settingsRepository.loadSettings();
    await this.launchAtStartupService.syncWithStoreValue(settings.launchAtStartup);
  }
}
